src/workerd/server/workerd.c++

```diff
@@ -848,9 +850,11 @@ public:
       }
     }
 
-    KJ_IF_SOME(c, config) {
-      Autogate::initAutogate(c);
-    }
+    // We'll fail at getConfig() if there are multiple top level Config objects.
+    // The error message says that you have to specify which config to use, but
+    // it's not clear that there is any mechanism to do that??
+    config::Config::Reader config = getConfig();
+    util::Autogate::initAutogate(config.getAutogates());
```

Comment:
This needs to check for the presence of the `autogates` field afaik. So you need to add a check for `hasAutogates` like the `initAutogate` method you removed did.

Reply:
I think this is not needed and `getAutogates` will always return an empty list if no `autogates` field is present.

Looking at the code, `workerd.capnp.h` has the definition of `getAutogates` which cleaned up a bit looks like:
```c++
List<autogate::Autogate>::Reader getAutogates() const {
  return PointerHelpers<List<autogate::Autogate>>::get(_reader.getPointerField(
      ::capnp::bounded<4>() * ::capnp::POINTERS));
```
Now `_reader` is a `capnp::StructReader` and `getPointerField` is defined [here](https://github.com/capnproto/capnproto/blob/v2/c%2B%2B/src/capnp/layout.h#L1127-L1135). It returns a `PointerReader` which records the offset where the data exists but doesn't dereference it.

`PointerHelpers<List<T>>::get(PointerReader)` is defined [here](https://github.com/capnproto/capnproto/blob/v2/c++/src/capnp/pointer-helpers.h?plain=1#L70-L73) and `defaultValue = nullptr` looks promising but the work is left to `List<T>::getFromPointer` and `List<T>::Reader` which are both defined in `list.h`. But then `List::getFromPointer` calls [`PointerReader.getList`](https://github.com/capnproto/capnproto/blob/v2/c++/src/capnp/layout.c++?plain=1#L2721) which does
```c++
pointer == nullptr ? &zero.pointer : pointer;
```
and then calls [`WireHelpers::readListPointer`](https://github.com/capnproto/capnproto/blob/v2/c++/src/capnp/layout.c++#L2269) the relevant part of which looks like:
```
    if (ref->isNull()) {
      if (defaultValue == nullptr ||
          reinterpret_cast<const WirePointer*>(defaultValue)->isNull()) {
        return ListReader(expectedElementSize);
      }
```
so in our case `ref` is `&zero.pointer` so probably `ref->isNull()` must return `true`, and `defaultValue` is `nullptr`, so we `return ListReader(expectedElementSize)`;. [The one argument `ListReader` constructor returns an empty list.](https://github.com/capnproto/capnproto/blob/v2/c++/src/capnp/layout.h#L762-L765)

So if I'm reading this correctly, in the case where the field is empty, it's filled in with a null pointer and when it is read capnp returns an empty list.

Reply:
The documentation for capnp says about struct Builders:
> getBar(): For primitives, returns the value. For composites, returns a Builder for the composite. If a composite field has not been initialized (i.e. this is the first time it has been accessed), it will be initialized to a copy of the field’s default value before returning.

But it doesn't comment on what `getBar` does or whether it is safe to call it without first checking for `hasBar` on `Readers`. My feeling is that if it was planning to segfault if you call `getBar` without first checking `hasBar` then this would be clearly declaimed. Though it would be nice for it to have some specific statement on the safety of `getBar` on missing fields...

Reply:
@kentonv Can you confirm this is how list defaults work? Is there an appropriate place in the documentation that spells out the safety requirements for using `StructReader.getKey()`?